include/splay.h

```diff
@@ -100,47 +100,132 @@ template<class V>
+            walkee(cur->data, state);
+            cur = cur->right;
+        } else {
+            // first descent into left subtree
+
+            // find right-most child in left tree
+            SplayNode<V> *rmc = cur->left;
+            while (rmc->right && rmc->right != cur)
+                rmc = rmc->right;
+
+            if (rmc->right == nullptr) {
+                // found the right most node in the left subtree:
+                // create a temporay thread back to cur to have
+                // a way back out of the subtree
+                rmc->right = cur;
+                // descent into left subtree
+                cur = cur->left;
+            } else {
+                // found thread in left subtree pointing back to cur:
+                // this means we moved back up a thread to cur
+                // in the last "descent into right" and completed
+                // the left subtree of cur.
+                // cut the temporary thread and call walkee on cur->data
+                rmc->right = nullptr;
+                walkee(cur->data, state);
+                // and descent into right subtree
+                cur = cur->right;
+            }
+        }
+    }
 }
 
 template<class V>
 SplayNode<V> const *
 SplayNode<V>::start() const
 {
-    if (left)
-        return left->start();
+    SplayNode<V> *cur = const_cast<SplayNode<V> *>(this);
+    while (cur->left)
+        cur = cur->left;
 
-    return this;
+    return cur;
 }
 
 template<class V>
 SplayNode<V> const *
 SplayNode<V>::finish() const
 {
-    if (right)
-        return right->finish();
+    SplayNode<V> *cur = const_cast<SplayNode<V> *>(this);
+    while (cur->right)
+        cur = cur->right;
 
-    return this;
+    return cur;
 }
 
 template<class V>
 void
 SplayNode<V>::destroy(SPLAYFREE * free_func)
 {
-    if (left)
-        left->destroy(free_func);
-
-    if (right)
-        right->destroy(free_func);
+    // to delete all nodes we use a modification of
+    // the Morris Traversal algorithm:
+    //
+    // Start with the top node as current node C.
+    //
+    // Repeat, until no more child nodes exist:
+    // a) If C has no left child, set C to the C->right
+    //    and delete the previous node C.
+    // b) If C has a left child C->left, find the right-most node RMN in the
+    //    left subtree of C.
+    //    - RMN can't have a right child!
+    //    - Instead of linking RMN->right to C as a way back up like in Morris
+    //      Traversal, we MOVE C->right to RMN->right.
+    //    - As C now has only a single child (C->left),
+    //      move C to C->left an delete the previous node C.
+    //
+    // Because we move the right subtree down into the left subtree and remove
+    // the parent from the tree, there is - in contrast to the Morris
+    // Traversal - no way up and no need to check for this loop case.
+    // We only need to avoid deleting the original top node (this) until all
+    // other nodes are deleted.
```

Comment:
The regular Morris Traveral is class O(n) too. But it is not really suited to delete nodes as far as I can see.
To avoid the need for extra data structures to keep track of multiple subtrees you have to delete "from the top".

The remove-method works quite similar:
- With the first "splay"-call it moves the element to delete into the top node.
- If the top node has no left child, the right child becomes the new top.
- If it has a left child, this becomes the new top node after another splay-call with the element to delete on the left subtree. Because this element has to be greater than all nodes in the left subtree, the new top of the left subtree doesn't have a right child after the splay-rotation (the right-most child was rotated to the top). Then the right subtree is inserted as right child of the top node of the left subtree (a.k.a. right-most-child) and the old top node can be removed.
The remove-algorithm can't be used for destroy because it needs a compare function for the node data, that's not known to destroy. It also would need a lot of unnecessary rotations and compares. My "destroy" variant combines the "removal of the top-node by moving the right subtree into the left subtree" method of remove with the "use the most-right-child in left subtree for linking" as a good new position for the right subtree. It even keeps the search order intact.
For destroy any empty child node in the left subtree would be sufficient as a new position for the right subtree, but because there is no knowledge about the depth of the individual branches, there is no best way to find the nearest empty child.

Reply:
> The remove-algorithm can't be used for destroy because ... It also would need a lot of unnecessary rotations and compares.

I agree with that part (at least).

> The regular Morris Traveral ... is not really suited to delete nodes as far as I can see.
> To avoid the need for extra data structures to keep track of multiple subtrees you have to delete "from the top".

I think that last statement is the crux of my question. I am probably missing something, but, when I look at where walkee() is called for  the regular Morris traversal, I do not see why we cannot essentially replace that call with `delete` as shown in the sketch below. This sketch is copy-paste of the relevant parts of your walk() implementation, with walkee() replaced with delete (and swapped around to come _after_ the `cur->right` is used, of course).

```C++
if (cur->left == nullptr) {
    const auto top = cur;
    cur = cur->right;
    delete top; // traversal will never see top again
} else {
    ...
    {
        // found thread in left subtree pointing back to cur
        rmc->right = nullptr;
        const auto top = cur;
        cur = cur->right;
        delete top; // traversal will never see top again
    }
}
```

I believe that if you replace `delete top` in the above with `walkee(top->data, state)`, you will get exactly the same traversal as in your walk() implementation. So the question is, why is it wrong to call delete in those two places? In other words, do we really revisit that `top` node at some point?

If we can use the above code, then we can delete the whole tree by calling the adjusted walk() with `delete` as a walkee function (essentially). We will need a wrapper to make regular walkee and delete calls compatible, but that is easy to do.

To avoid misunderstanding: Even if the regular traversal algorithm _can_ be reused for destroying the tree, I do not insist on that code reuse if you see enough _value_ in that custom destroy() implementation. In that case, we just need to document what that value. That is why I asked about its advantages over the regular traversal. If the regular traversal algorithm cannot be reused for destroying the tree, then we have to use a different implementation for destroy(), of course.